locator/tablets.cc

```diff
@@ -523,6 +553,34 @@ class tablet_effective_replication_map : public effective_replication_map {
         return result;
     }
 
+    virtual dht::token_range_vector get_ranges(inet_address ep) const override {
+        const auto& tm = *_tmptr;
+        const auto& sorted_tokens = tm.sorted_tokens();
+        if (sorted_tokens.empty()) {
+            on_internal_error(rslogger, "Token metadata is empty");
+        }
+
+        dht::token_range_vector ret;
+
+        auto prev_tok = sorted_tokens.back();
+        for (const auto& tok : sorted_tokens) {
```

Comment:
It's wrong to use sorted_tokens here, they're only relevant for vnodes. Use get_tablet_map() to get the table-specific layout.

You can use first_tablet/next_tablet to iterate. Ideally you'd use for_each_tablet() but that returns a future, so add a FIXME to change get_ranges() to return a future.

Reply:
btw, you won't need to use insert_token_range_to_sorted_container_while_unwrapping since tablets don't wrap around the ring (end of tablet always > start of tablet).

Reply:
Sure, corrected the implementation as follows:

```
+    virtual dht::token_range_vector get_ranges(inet_address ep) const override {
+        dht::token_range_vector ret;
+
+        auto&& tablets = get_tablet_map();
+        tablets.for_each_tablet([&](locator::tablet_id id, const tablet_info& info) -> future<> {
+            auto endpoints = get_natural_endpoints(tablets.get_last_token(id));
+            auto should_add_range = std::find(std::begin(endpoints), std::end(endpoints), ep) != std::end(endpoints);
+
+            if (should_add_range) {
+                ret.push_back(tablets.get_token_range(id));
+            }
+
+            return make_ready_future<>();
+        }).get();
+
+        return ret;
+    }
```

To enable usage of the new implementation I had to adjust its caller - `storage_service::effective_ownership()`.

It uses sorted tokens to describe the ownership:
`const auto token_ownership = dht::token::describe_ownership(tm.sorted_tokens())`

I had to create `token_ownership` in a different way. Instead of using `sorted_tokens()` I had to get tokens vector from tablet map. Right now I am extending the tests to check if it works correctly in all cases - the basic case with single node and RF=1 passed.